tests/ui/pin-ergonomics/pattern-matching-deref-pattern.rs

```diff
@@ -0,0 +1,91 @@
+//@ revisions: pin_ergonomics normal
+//@ edition:2024
+//@ check-pass
+// //@[normal] check-pass
+// //@[pin_ergonomics] rustc-env:RUSTC_LOG=rustc_hir_typeck::expr_use_visitor=DEBUG
+#![cfg_attr(pin_ergonomics, feature(pin_ergonomics))]
+#![feature(deref_patterns)]
+#![allow(incomplete_features)]
+
+// This test verifies that the `pin_ergonomics` feature works well
+// together with the `deref_patterns` feature.
+
+use std::pin::Pin;
+
+struct Foo<T, U> {
+    x: T,
+    y: U,
+}
+
+struct Bar<T, U>(T, U);
+
+enum Baz<T, U> {
+    Foo(T, U),
+    Bar { x: T, y: U },
+}
+
+fn foo_mut<T: Unpin, U: Unpin>(foo: Pin<&mut Foo<T, U>>) {
+    let Foo { .. } = foo;
+    let Pin { .. } = foo;
+    let _ = || {
+        let Foo { .. } = foo;
+        let Pin { .. } = foo;
+    };
+
+    #[cfg(pin_ergonomics)]
+    let Foo { x, y } = foo;
```

Comment:
Could you test the type of the bindings like you do in the other tests? Could you also remove the `cfg` so we can see what happens with just `deref_patterns` (I _think_ it should work, but seeing the error is good too).

Reply:
I uncommented this, but got an error of `cannot move out of a shared reference`. Seems that `x` and `y` have by-value bindings. Do you know if this behaviour is expected?